### ISSUE TITLE
Modules management improvements

### ISSUE THREAD

**Issue by rjarry** [sysrepo/sysrepo#1917]
Hi Michal,

In the legacy branch, `sysrepoctl` was able to re-install an already installed module. This is not possible anymore. `sysrepoctl --install` only works if the module is *not* installed. If the module is already installed with the same revision (or even with a different revision), it fails.

And `sysrepoctl --update` does not work when the revisions did not change. This is quite tedious when working as we need to add a revision every time we change small things.

Would you consider adding a `--force` option to both `sysrepoctl --install` and `sysrepoctl --update` (and also maybe `sysrepoctl --remove`) ?

Ideally, there should be a way to do `sysrepoctl --force --install-or-update my-module.yang` which would do what's required. This would avoid overly complex logic in shell scripts. I figure the `--force` and the `--install-or-update` options would be reflected in the C API to use without `sysrepoctl`.

If that seems ok for you, I can try to do a draft patch.

Thanks in advance.

**Comment by michalvasko** [comment]
Hi,
`sysrepoctl --install` for an already installed module (in the same revision) prints an error but should actually return 0, which was changed precisely having scripts in mind. Installing an already installed module in a different revision must fail because there can always be only one revision of a module implemented (YANG requirement).

`sysrepoctl --update` works the way you described but I think it is already quite lenient. It could require your modules to conform to the [update rules](https://tools.ietf.org/html/rfc7950#section-11) but all of that was reduced to only requiring a new revision. Also, I believe what you are describing as tedious was actually the exact intention of the specification. It makes sense for deployment because it forces you to never define 2 modules that are not completely similar but have the same name/revision.

Adding any options such as `--force` would require non-backwards compatible changes (all these operations are simply API functions). Generally, I am not against that if it is required but that is not the case. I would need much more details as to why this is required.

What exactly do you mean by overly complex logic? When you are installing new modules I believe you should care that a similar module is already installed, especially if it is in a different revision.

It seems to me you want to migrate your scripts that you used in the old *sysrepo* with minimal effort, which is understandable, but for this purpose correctness will never be sacrificed. So, please try to present your use-case better. Also, I would have nothing against changing `sysrepoctl --list` output to allow for some additional parameters/formatting so that it can be processed in scripts more easily. That may be useful for detailed examination of current *sysrepo* modules so that the scripts can adjust their commands accordingly.

Regards,
Michal

**Comment by rjarry** [comment]
Hi Michal,

I have absolutely no issue if reworking all our scripts from scratch if that is required. My main goal is to make it work. Whatever the cost :)

Let me clarify our use case: we have approximately 70 YANG modules. For modularity purposes, we have a base module which contains almost nothing but two root containers. All other modules `augment` the base module with additional data and/or rpcs. This means a lot of dependencies between these modules.

When working in a development environment, every time we make a slight modification in one of the modules, we need to re-install(update) the module in sysrepo.

Until now (with the *old sysrepo*), we managed that with `sysrepoctl` commands. Since module update was flaky, we needed to completely uninstall all our modules first, and then reinstall them all. Only uninstalling our base module was not possible because of dependencies.

Now that there is an `--update` option, I tried to use it but it would mean adding a new (dummy) revision to modules for each little change made during development. I understand the need for revisions to identify changes in modules, but this has been designed with "production" environment in mind. This is not practical when in a development environment where we make (temporary and minor) modifications to YANG modules multiple times per hour.

In any case, I gave up with this `--update` option and tried to uninstall/install all modules, like it was done before. However, it does not work either. The uninstallation fails because of dependency problems. It seems I am stuck :)

To sum-up our need: We need to be able to do *micro* updates of our modules during the development phase without requiring a new YANG revision every time. If that means uninstalling everything every time, we can live with that.

I am willing to share our complete set of YANG modules and more details about the exact errors if you want.

Thanks in advance for your help.

As a side note:

> `sysrepoctl --install` for an already installed module (in the same revision) prints an error but should actually return 0,

This is very misleading. If there is no problem, the error should *not* be printed. I see that when running `make install` multiple times in Netopeer2:

```
~# make install
...
-- Installing sysrepo modules...
[ERR]: Module "ietf-netconf-acm" not scheduled for deletion.
[ERR]: Module "ietf-netconf-acm" is already in sysrepo.
[ERR]: Module "ietf-netconf@2013-09-29" already installed.
sysrepoctl error: Failed to update module "/root/sshfs_root/Netopeer2/server/../modules/ietf-netconf@2013-09-29.yang" (Item already exists)
-- Generating a new RSA host key "genkey"...
-- Merging default server listen configuration...
```

**Comment by michalvasko** [comment]
Hi,
okay, fair enough. If we are talking only about development then I understand and can tell you how I do it. Every time I need to change some modules or, most often in my case, make *sysrepo* work again after it crashed, I simply run `make sr_clean`, which completely resets everything (modules, data). Then you can normally install all the modules (using a script) again. If this is not suitable for you then we can think about something better but I would prefer a similar out-of-sight "hidden" solution because it is meant only for development.

As for *netopeer2* install scripts, you are absolutely right, they work only if you do not have the modules installed. I should improve them but there are simply much more important things for me to do right now. Also, I consider installing an already installed module an error so I think the message is correct (and the API function returns an error, only *sysrepoctl* hides it because someone requested it) and you should ideally check this situation before installing the module.

Regards,
Michal

**Comment by rjarry** [comment]
I have made a draft patch to address the issues I mentioned here. Could you please tell me what you think about it. And maybe merge the pull request if that looks ok to you :)

Thanks in advance.

**Comment by michalvasko** [comment]
Hi,
for future, it would be better to wait until a conclusion is reached, which can save you your work. Based on your changes it seems to me you mostly want to ignore all the errors. That can be achieved by simply ignoring the return value of `sysrepoctl` and turn its logging off (`-v0`) so I think those changes are pointless. The exception is module update and I am not sure what to do about that one. Like I said, it is already very permissive so I would prefer to keep it the way it is. As you correctly observed, it is not really meant for development but rather as a safe way of updating a module without having to backup its data in deployment.

In short, I think there is little benefit in your changes to justify adding 3 functions into API. So, unless you come up with less invasive changes (also, API functions to be used only for development?), I will keep things the way they are, sorry.

Regards,
Michal

**Comment by rjarry** [comment]
Hi Michal,

> Based on your changes it seems to me you mostly want to ignore all the errors. That can be achieved by simply ignoring the return value of sysrepoctl and turn its logging off (-v0) so I think those changes are pointless.

I am not ignoring *all* errors. Only ignoring certain ones. Ignoring all errors would not be desirable. I want to be able to see if the module cannot be installed because of some real problem (missing dependency, parsing error, etc.). I only want to ignore the `module already installed`, `revision is the same` or `module already removed` errors (I do not consider these exactly "errors").

Ideally, these operations should be idempotent (i.e.: running `sr_module_install_force()` twice with the same module should not produce any warning). My patch achieves exactly that.

I am willing to rework it if necessary. In any case, we need something of the sort. I'll keep it in our local changes in the meantime.

Thanks for taking the time to look at it.

**Comment by michalvasko** [comment]
Hi,
okay, I understand. But generally there is no reason to, for example, install a module that has already been installed. It is only convenient for cases when it does not matter. At the same time it leaves the use-cases when it does matter prone to errors because it would not be detected. And after enough experience with YANG and the rest, I now try to always favor correctness over convenience.

Naturally, it would be best if somehow both cases could be covered reasonably. We can leave this for now and maybe in future you or I will come up with something. For example, adding functionality into `sysrepoctl` that can tell whether a particular module is installed would be quite simple, would not affect API at all, and should cover your problems with installing/removing modules.

Regards,
Michal
